Unbounded timeline selection

danielgavrilov opened the pull request:
Notion issue: https://www.notion.so/owid/Option-to-set-timeline-end-to-latest-available-year-date-1d2c44e0db054b2b9c518bcaa58e835e

When the **start/end year** of the timeline is set **beyond the minimum/maximum year**, the timeline becomes **unbounded** in that direction.
 
(Note that due to rounding, it's still possible to set the timeline to the latest year when the cursor is in between the latest and second latest year. This is intentional. We should ideally have some clever UI to signal this...)

**Issues**:

- ~~The timeline **is not unbounded for single-year line charts and single-year scatter plots** and won't be unbounded for any other charts that select a single year but also support a range. 
    This is because bounding is implemented with `undefined`, which lacks a direction, and charts can assume different directions. All charts handle an `undefined` end year as "set to maximum", so that is straightforward. But an `undefined` start year is interpreted as maximum on Maps, and minimum on Line charts. Possibly in the future we could add directionality with `-Infinity` and `+Infinity`, but that isn't handled well now.~~

- Adds tests for URL parameter handling, but not for the timeline, which is more problematic. I didn't test the timeline because I expect it would be tricky to simulate all the possible interactions and take lots of time.

danielgavrilov commented on charts/__tests__/ComparisonLineGenerator.test.ts:
```diff
@@ -1,4 +1,4 @@
-#! /usr/bin/env jest
+#! /usr/bin/env yarn jest
```
@breck7 I think this way it uses `node_modules` rather than the globally installed `jest`. But let me know if there are possibly issues defining it like this

breck7 replied:
This way with yarn is better, thanks!

breck7 commented on charts/Controls.tsx:
```diff
@@ -442,8 +442,8 @@ class TimelineControl extends React.Component<{ chart: ChartConfig }> {
                     onTargetChange={this.onChartTargetChange}
-                    startYear={chart.scatter.startYear}
-                    endYear={chart.scatter.endYear}
+                    startYear={chart.props.minTime}
```
Nice to generalize this more!

breck7 commented on charts/HTMLTimeline.tsx:
```diff
@@ -310,6 +332,12 @@ export class Timeline extends React.Component<TimelineProps> {
 
     // Allow proper dragging behavior even if mouse leaves timeline area
     componentDidMount() {
+        console.log("componentDidMount")
```
nit: I'm guessing left in by accident?

danielgavrilov commented:
I took a bit of a dive into this. It would be great if you can review @breck7.

Summary of code changes:

- Implemented directional bounds, so that e.g. `targetYear` can be both unbounded on the **right** (always picking the maximum available year) or unbounded to the **left** (picking the minimum)
  - The direction is implemented with `+Infinity` and `-Infinity`
  - JSON doesn't handle `Infinity` so `minTime`, `maxTime` and `map.targetYear` in the database are stored as `"latest"` and `"earliest"` if they are unbounded, otherwise they stay integers as they are now. On load, they are reserialized into valid props, so client-side types are not affected.
  - The same string formatting applies to single-year params in URLs, e.g. can have `year=latest` (for maps) or `time=latest` for single-year line charts (bar charts).
  - For intervals we don't need directionality on start/end, so they are implemented as suggested in #349: `[int]..[int]` where `[int]` is ommited if unboounded.
- Created `TimeBound.ts` utilities to deal with this unbounded mess.
- ⭐️ Created an **abstract `ChartTransform` class** to standardize the `timelineYears`, `startYear` and `endYear` properties.
- Added tests for JSON & URL serialization & deserialization of time properties.
- Also started using a bit of docstrings to document things – VSCode displays them automatically on hover, which I think is super useful as the codebase grows.

Summary of UI changes:
- You can now drag timeline handles beyond start/end to set them as "unbounded"
- You can erase the timeline fields in Grapher Admin to set them as "unbounded"

MarcelGerber commented on charts/TimeBounds.ts:
```diff
@@ -0,0 +1,129 @@
+    [minTime, maxTime]: [Time, Time],
+    time: Time
+): TimeBound {
+    if (time <= minTime) return TimeBoundValue.unboundedLeft
+    if (time >= maxTime) return TimeBoundValue.unboundedRight
```
Do we really jump to "latest" if the value is equal (`=`) to `maxTime`, or only if bigger (`>`)?

danielgavrilov replied:
Yes, the timeline only goes up to `maxTime` as I implemented it, so it has to be `=`.

It can be set to the latest year without becoming unbounded by setting it to just before `maxTime`, and then it gets rounded up.

MarcelGerber commented on charts/TimeBounds.ts:
```diff
@@ -0,0 +1,129 @@
+    return Math.min(maxTime, Math.max(minTime, bound))
+}
+
+export function getTimeFromTimes(
```
I find the name here... not so great

danielgavrilov replied:
Haha nope! It's hard to name these things. The differences are quite subtle, I didn't manage to come up with a good one – let me know if you have suggestions

I opted for docstrings to clear up some parts, but that's definitely not ideal.

MarcelGerber replied:
Not a direct suggestion, but `findClosestYear` is a way better name and is similar to what this does. So, maybe `getClosestTime`?

danielgavrilov replied:
Good point, it's using `findClosestYear` so it should have `closest` in the name. Now changed, thanks!

MarcelGerber commented on charts/DiscreteBarTransform.ts:
```diff
@@ -41,16 +35,14 @@ export class DiscreteBarTransform implements IChartTransform {
-        const yearsUniq = flatten(
-            this.primaryDimensions.map(dim => dim.variable.yearsUniq)
+    @computed get timelineYears(): Time[] {
+        return sortedUniq(
+            sortBy(
+                flatten(
+                    this.primaryDimensions.map(dim => dim.variable.yearsUniq)
+                )
+            )
```
nit: You could be more consistent with how these are implemented in all the transforms :)
Sometimes it's `map, flatten, sortBy, sortedUniq`, other times its `map, union, sortBy, sortedUniq` (`union` already does `uniq`!). Btw, `map + flatten` is just `flatMap`.

MarcelGerber commented on charts/ChartUrl.ts:
```diff
@@ -234,16 +234,20 @@ export class ChartUrl implements ObservableUrl {
 
         const time = params.time
-        if (time !== undefined) {
+        if (time) {
```
This won't work for `"0"`.

danielgavrilov replied:
I think it will? `""` is falsy, but `"0"` is truthy.

MarcelGerber replied:
You're right 😆

MarcelGerber commented on charts/HTMLTimeline.tsx:
```diff
@@ -52,102 +58,91 @@ export class Timeline extends React.Component<TimelineProps> {
-            maxYear,
-            Math.max(minYear, Math.max(startYearInput, endYearInput))
-        )
+    getYearRound(bound: TimeBound, defaultValue: TimeBound): TimeBound {
```
Well, it's not really rounding...

danielgavrilov replied:
Yes, I had actually renamed this method to something else. And then came back to Round because it felt most appropriate.

The timeline feeds decimals into this function and it returns the closest value in an array... not quite round, but not sure what to call it.

MarcelGerber replied:
Ah alright, that makes more sense. I guess I was confused because I thought `TimeBound` can only ever be an integer (or infinity), but right, it can be everything.
But even then, it isn't intuitive that this can still return +-Infinity. Judging from the name one would think this would find some year from the timeline.

danielgavrilov replied:
Also renamed this to `{start,end}YearClosest`, feels righter...

danielgavrilov commented on charts/ChartTransform.ts:
```diff
@@ -0,0 +1,103 @@
+import { ChartConfig } from "./ChartConfig"
+
+export interface IChartTransform {
+    isValidConfig: boolean
```
I just kept it as it was 🙂 The transforms already have that property – I'm not sure if it should be a getter, but probably best to not refactor that in this PR.

danielgavrilov commented on test/utils.ts:
```diff
@@ -0,0 +1,25 @@
+}
+
+export function setupChart(
+    id: 677 | 792,
+    varId: 3512 | 104402,
```
It was `number`, but I always look up which IDs we have in the `fixtures` folder, and I thought it's more convenient if VSCode can just show that in the tooltip as you're writing the function. Can definitely change it to `number` if there are downsides!

MarcelGerber replied:
Alright! No, I can't think of any immediate downsides, just that we have to keep this in sync (but since TS will complain otherwise, that should be a given)

danielgavrilov commented on charts/__tests__/ChartUrl.ts:
```diff
@@ -0,0 +1,176 @@
+import { ChartUrl, ChartQueryParams } from "../ChartUrl"
+import { MapConfigProps } from "charts/MapConfig"
+
+type TimeDomain = [Time, Time]
```
Good idea – I've now created `TimeBounds`.

These types are very confusing because I gave them a bad name (I couldn't come up with a better one) and because they are the same TypeScript type – `number`. 

`Time` is supposed to represent a number with a finite value, while `TimeBound` can contain `Infinity`. But I couldn't figure out if there is a way to enforce that a bit better...

breck7 commented on charts/ChartConfig.tsx:
```diff
@@ -172,11 +181,11 @@ export class ChartConfigProps {
 
-    @observable.ref timelineMinTime?: number = undefined
-    @observable.ref timelineMaxTime?: number = undefined
+    @observable.ref timelineMinTime?: Time = undefined
```
Do we need a minTime and a timelineMinTime? Do we have 2 time filters/

danielgavrilov replied:
Not great naming... but one of them sets the start & end handles in the timeline, and the other limits what the minimum/maximum possible value can be selected (so you can hide some years from the timeline)

danielgavrilov replied:
I think it's good to start writing some docstrings around these properties, since they are pretty stable given they are in the database, for better or worse...

breck7 commented on charts/Controls.tsx:
```diff
@@ -442,8 +443,8 @@ class TimelineControl extends React.Component<{ chart: ChartConfig }> {
                 <Timeline
```
Awesome generalization. Maybe at some point we could consolidate these almost identical Timeline elements into 1 and only change the different properties in each switch.

breck7 commented on charts/DataTableTransform.ts:
```diff
@@ -148,17 +149,28 @@ export class DataTableTransform {
     }
 
+    @computed get minYear(): Time {
+        return this.chart.data.minYear ?? 1990
```
nit: any reason why the defaults for this are 1990 and 2020, and 1900 and 2000 for the others? Maybe setting them as a constant like ARBITRARY_MIN_YEAR might be clearer.

breck7 commented on charts/DiscreteBarTransform.ts:
```diff
@@ -4,27 +4,21 @@ import {
-        this.chart = chart
-    }
-
+export class DiscreteBarTransform extends ChartTransform {
```
Nice. I think for core things like transforms using inheritance makes sense. (I was a little surprised that React discourages inheritance use so much, but at least for things other than components we could benefit from using it more).

danielgavrilov replied:
Now that I think about it, classes don't compose well, and we will likely eventually move away from using them in transforms.

Maybe a good example is that I didn't extend this class for `MapData` and `DataTableTransform` because the interfaces they provide are a little bit different. But they could still maybe share some subset of logic in the future.

breck7 commented on charts/HTMLTimeline.tsx:
```diff
@@ -52,102 +58,91 @@ export class Timeline extends React.Component<TimelineProps> {
+    @observable startYearRaw: TimeBound = TimeBoundValue.unboundedLeft
+    @observable endYearRaw: TimeBound = TimeBoundValue.unboundedRight
+
+    @computed get startYear(): TimeBound {
```
Nit: Can we make privates private? Mildly helpful for readability and later refactoring.

danielgavrilov replied:
Given that this is a React component, and they are instantiated in JSX, isn't privacy assumed for all their properties/methods?

breck7 replied:
Ah, fair point. I am just used to working with pure TS more where explicit privates are very valuable during refactoring.

breck7 commented on charts/LineChartTransform.ts:
```diff
@@ -119,42 +106,12 @@ export class LineChartTransform implements IChartTransform {
-            this.maxTimelineYear
-        )
-    }
-
```
+1 for removing code

danielgavrilov commented:
> The only thing I could come up with is hitting play and waiting for the animation to finish. When it does the end time will not be unbounded.

Yes I'm not sure if it should become unbounded at the end of playing... I intentionally made it so that it can only be made unbounded by user interaction when the handle is dragged to `>= maxTime`.

This means you can drag the handle to just _before_ the end of the timeline, and due to rounding it will be set to the latest year without being unbounded.

I think it's good to have the option to set the latest year (bounded), but the UI definitely doesn't make it clear that both states are possible. 

But since this feature is a small detail that I think would only be used internally with changing datasets, I hope it doesn't matter to most users whether something is bounded/unbounded.

danielgavrilov commented:
Thanks a lot for reviewing @breck7 and @MarcelGerber! 

I will merge and deploy this as it is so that we can stop shifting the slider manually on 50 charts every day. I made some small changes to address the comments, but let me know if it could be improved still.
